projects/batfish/src/main/java/org/batfish/job/BatfishJobExecutor.java

```diff
@@ -41,105 +50,143 @@ public BatfishJobExecutor(
-              if (!_haltOnProcessingError) {
-                result.applyTo(output, _logger, answerElement);
-              }
-            }
-          }
-        } else {
-          continue;
-        }
-      }
-      if (!futures.isEmpty()) {
+
+    for (int i = 0; i < jobs.size(); i++) {
+      try {
+        // getting the next completed task
+        Future<JobResultT> currentDoneFuture = completionService.take();
+        updateJobsStats();
+
+        JobResultT result = null;
         try {
-          Thread.sleep(JOB_POLLING_PERIOD_MS);
-        } catch (InterruptedException e) {
-          throw new BatfishException("interrupted while sleeping", e);
+          // getting the result of the job
+          result = currentDoneFuture.get();
+        } catch (InterruptedException | ExecutionException e) {
+          throw new BatfishException("Error executing job", e);
+        }
+        boolean jobResultError = handleJobResult(result, output, answerElement, failureCauses);
+        if (jobResultError) {
+          processingError = true;
         }
+
+      } catch (InterruptedException e) {
+        throw new BatfishException("interrupted while waiting for jobs to complete", e);
       }
     }
     pool.shutdown();
     if (processingError) {
-      int numJobs = jobs.size();
-      int numFailed = numJobs - failureCauses.size();
-      int numSucceeded = numJobs - numFailed;
-      if (!_logger.isActive(BatfishLogger.LEVEL_INFO)) {
-        _logger.infof("%d jobs succeeded; %d jobs failed\n", numSucceeded, numFailed);
-      }
-      if (_haltOnProcessingError) {
-        throw new CompositeBatfishException(
-            new BatfishException("Fatal exception due to failure of at least one job"),
-            failureCauses);
-      }
+      handleProcessingError(jobs, failureCauses);
     } else if (!_logger.isActive(BatfishLogger.LEVEL_INFO)) {
       _logger.info("All jobs executed successfully\n");
     }
   }
+
+  ExecutorService getExecutorService() {
+    ExecutorService pool;
+    if (!_settings.getSequential()) {
+      int maxConcurrentThreads = Runtime.getRuntime().availableProcessors();
+      int numConcurrentThreads = Math.min(maxConcurrentThreads, _settings.getJobs());
+      pool = Executors.newFixedThreadPool(numConcurrentThreads);
+    } else {
+      pool = Executors.newSingleThreadExecutor();
+    }
+    return pool;
+  }
+
+  /**
+   * Handles the Job result appropriately in case of success/error, and updates logs
+   *
+   * @param result Result of Job
+   * @param output Output comprising results of all jobs
+   * @param answerElement Answer Element for the execution of jobs
+   * @param failureCauses List of failure causes for all jobs
+   * @return True if there is a processing error else False
+   */
+  boolean handleJobResult(
+      JobResultT result,
+      OutputT output,
+      AnswerElementT answerElement,
+      List<BatfishException> failureCauses) {
+    boolean processingError = false;
```

Comment:
drop this useless local variable; just `return false` or `true` in the places you need to return.

Reply:
done